### PR TITLE
En  6314/activate hardfork trigger

### PR DESCRIPTION
**Pull request by sasurobert**
Activate hardfork trigger.
Trigger is called by a trusted node, which broadcasts the data through heartbeat. Every node can set if they accept or not this message - it is not enforced. Every validator can set the key from where they accept this message.
The trigger is registered and act upon depending on the epoch. If it is for a past epoch the node instantly starts the Export process.
If the trigger is for a new epoch, the message is registered, and the trigger handler waits for the EPOCH change confirmation - coming from shardChain(when metablock with epoch is finalized) and metaChain trigger(when the block following the epoch start metablock is commited).

### CODE REVIEW

File: update/trigger/trigger.go [hunk 1]
```diff
@@ -69,30 +110,88 @@ func (t *trigger) getCurrentUnixTime() int64 {
...
+		log.Error("error while creating export handler", "error", err)
+		return
+	}
+
+	err = exportHandler.ExportAll(t.epoch)
+	if err != nil {
+		log.Error("error while exporting data", "error", err)
+		return
+	}
+
```
Comment (iulianpascalau) [inline update/trigger/trigger.go:171]: also here, stating that the hardfork ended in success and a node close is pending

File: update/trigger/trigger.go [hunk 2]
```diff
@@ -106,50 +205,68 @@ func (t *trigger) TriggerReceived(originalPayload []byte, data []byte, pkBytes [
...
 	for _, handler := range t.triggerHandlers {
-		go handler()
+		go handler(epoch)
 	}
 	t.mutTriggerHandlers.RUnlock()
 }
 
 // RegisterHandler will add a trigger event handler to the list
-func (t *trigger) RegisterHandler(handler func()) error {
+func (t *trigger) RegisterHandler(handler func(epoch uint32)) error {
```
Comment (iulianpascalau) [inline update/trigger/trigger.go:269]: do we still need these? (RegisterHandler, callAddedDataHandlers, associated mutex and triggerHandlers)?

Comment (sasurobert) [reply update/trigger/trigger.go:269]: maybe not.

File: cmd/node/main.go [hunk 3]
```diff
@@ -1823,6 +1846,94 @@ func getConsensusGroupSize(nodesConfig *sharding.NodesSetup, shardCoordinator sh
...
+
+	atArgumentParser := vmcommon.NewAtArgumentParser()
+	argTrigger := trigger.ArgHardforkTrigger{
+		TriggerPubKeyBytes:     triggerPubKeyBytes,
+		SelfPubKeyBytes:        selfPubKeyBytes,
+		Enabled:                config.Hardfork.EnableTrigger,
+		EnabledAuthenticated:   config.Hardfork.EnableTriggerFromP2P,
+		ArgumentParser:         atArgumentParser,
+		EpochProvider:          process.EpochStartTrigger,
+		ExportFactoryHandler:   hardForkExportFactory,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/main.go:1925]: there is no need to make another component which registers and that one knows about the factory and the export - that is a component of 20 lines of code.

File: cmd/node/main.go [hunk 4]
```diff
@@ -1823,6 +1846,94 @@ func getConsensusGroupSize(nodesConfig *sharding.NodesSetup, shardCoordinator sh
...
+	if err != nil {
+		return nil, err
+	}
+	triggerPubKeyBytes, err := stateComponents.ValidatorPubkeyConverter.Decode(config.Hardfork.PublicKeyToListenFrom)
+	if err != nil {
+		return nil, fmt.Errorf("%w while decoding HardforkConfig.PublicKeyToListenFrom", err)
+	}
+
+	accountsDBs := make(map[state.AccountsDbIdentifier]state.AccountsAdapter)
+	accountsDBs[state.UserAccountsState] = stateComponents.AccountsAdapter
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/main.go:1877]: these accounts adapters are needed for you to export and share your own data - new tries are created in the export folder

File: epochStart/notifier/epochStartSubscriptionHandler.go [hunk 5]
```diff
@@ -84,6 +87,26 @@ func (essh *epochStartSubscriptionHandler) NotifyAllPrepare(metaHdr data.HeaderH
...
+	essh.mutEpochStartHandler.Lock()
+	essh.epochFinalizedHandler = append(essh.epochFinalizedHandler, handler)
+	essh.mutEpochStartHandler.Unlock()
+}
+
+// NotifyEpochChangeConfirmed will call all the subscribed clients to notify them that an epoch change is confirmed
+func (essh *epochStartSubscriptionHandler) NotifyEpochChangeConfirmed(epoch uint32) {
+	essh.mutEpochStartHandler.Lock()
+	for _, handler := range essh.epochFinalizedHandler {
+		handler(epoch)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/notifier/epochStartSubscriptionHandler.go:105]: done

File: integrationTests/multiShard/transaction/interceptedResolvedBulkTx/interceptedResolvedBulkTx_test.go [hunk 6]
```diff
@@ -64,6 +64,7 @@ func TestNode_InterceptorBulkTxsSentFromSameShardShouldRemainInSenderShard(t *te
 	transactionValue := big.NewInt(1)
 	senderPrivateKeys := []crypto.PrivateKey{nodes[idxSender].OwnAccount.SkTxSign}
 	integrationTests.CreateMintingForSenders(nodes, shardId, senderPrivateKeys, balanceValue)
+	_ = logger.SetLogLevel("*:TRACE")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/multiShard/transaction/interceptedResolvedBulkTx/interceptedResolvedBulkTx_test.go:67]: done

File: update/trigger/trigger.go [hunk 7]
```diff
@@ -69,30 +110,88 @@ func (t *trigger) getCurrentUnixTime() int64 {
...
+func (t *trigger) exportAll() {
+	t.mutTriggered.Lock()
+	defer t.mutTriggered.Unlock()
+
+	exportHandler, err := t.exportFactoryHandler.Create()
+	if err != nil {
+		log.Error("error while creating export handler", "error", err)
+		return
+	}
+
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/trigger/trigger.go:165]: done

File: update/trigger/trigger.go (hunk 1 above)
Comment (sasurobert) [reply update/trigger/trigger.go:171]: done

File: update/mock/epochStartNotifierStub.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+func (esnm *EpochStartNotifierStub) NotifyEpochChangeConfirmed(epoch uint32) {
+	if esnm.NotifyEpochChangeConfirmedCalled != nil {
+		esnm.NotifyEpochChangeConfirmedCalled(epoch)
+	}
+}
+
+// RegisterHandler -
+func (esnm *EpochStartNotifierStub) RegisterHandler(handler epochStart.ActionHandler) {
+	if esnm.RegisterHandlerCalled != nil {
+		esnm.RegisterHandlerCalled(handler)
```
Comment (SebastianMarian) [inline update/mock/epochStartNotifierStub.go:36]: Forgot to add return here ?

File: update/mock/epochStartNotifierStub.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	esnm.epochStartHdls = append(esnm.epochStartHdls, handler)
+}
+
+// UnregisterHandler -
+func (esnm *EpochStartNotifierStub) UnregisterHandler(handler epochStart.ActionHandler) {
+	if esnm.UnregisterHandlerCalled != nil {
+		esnm.UnregisterHandlerCalled(handler)
```
Comment (SebastianMarian) [inline update/mock/epochStartNotifierStub.go:45]: Forgot to add return here ?

File: update/mock/epochStartNotifierStub.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+			esnm.epochStartHdls = append(esnm.epochStartHdls[:i], esnm.epochStartHdls[i+1:]...)
+			break
+		}
+	}
+}
+
+// NotifyAllPrepare -
+func (esnm *EpochStartNotifierStub) NotifyAllPrepare(metaHdr data.HeaderHandler, body data.BodyHandler) {
+	if esnm.NotifyAllPrepareCalled != nil {
+		esnm.NotifyAllPrepareCalled(metaHdr, body)
```
Comment (SebastianMarian) [inline update/mock/epochStartNotifierStub.go:59]: Forgot to add return here ?

File: update/mock/epochStartNotifierStub.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+	for _, hdl := range esnm.epochStartHdls {
+		hdl.EpochStartPrepare(metaHdr, body)
+	}
+}
+
+// NotifyAll -
+func (esnm *EpochStartNotifierStub) NotifyAll(hdr data.HeaderHandler) {
+	if esnm.NotifyAllCalled != nil {
+		esnm.NotifyAllCalled(hdr)
```
Comment (SebastianMarian) [inline update/mock/epochStartNotifierStub.go:70]: Forgot to add return here ?

File: update/trigger/trigger.go [hunk 12]
```diff
@@ -49,18 +69,39 @@ func NewTrigger(arg ArgHardforkTrigger) (*trigger, error) {
...
 		return nil, fmt.Errorf("%w self public key bytes length is 0", update.ErrInvalidValue)
 	}
+	if check.IfNil(arg.ArgumentParser) {
+		return nil, update.ErrNilArgumentParser
+	}
+	if check.IfNil(arg.EpochProvider) {
+		return nil, update.ErrNilEpochHandler
+	}
+	if check.IfNil(arg.ExportFactoryHandler) {
+		return nil, update.ErrNilExportHandlerFactory
```
Comment (SebastianMarian) [inline update/trigger/trigger.go:79]: ErrNilExporFactoryHandler

File: update/trigger/trigger.go [hunk 13]
```diff
@@ -69,30 +110,90 @@ func (t *trigger) getCurrentUnixTime() int64 {
...
 
-// Trigger will start of the hardfork process
-func (t *trigger) Trigger() error {
+func (t *trigger) epochConfirmed(epoch uint32) {
 	if !t.enabled {
-		return update.ErrTriggerNotEnabled
+		return
+	}
+
+	t.mutTriggered.Lock()
```
Comment (SebastianMarian) [inline update/trigger/trigger.go:118]: RLock/RUnlock ?

File: update/trigger/trigger.go [hunk 14]
```diff
@@ -69,30 +110,90 @@ func (t *trigger) getCurrentUnixTime() int64 {
...
+	if !isTriggered || triggeredEpoch > epoch {
+		return
 	}
 
-	return t.doTrigger(nil)
+	t.exportAll()
 }
 
-func (t *trigger) doTrigger(payload []byte) error {
+// Trigger will start of the hardfork process
```
Comment (SebastianMarian) [inline update/trigger/trigger.go:130]: without "of"

File: update/trigger/trigger.go [hunk 15]
```diff
@@ -106,50 +207,68 @@ func (t *trigger) TriggerReceived(originalPayload []byte, data []byte, pkBytes [
...
+
+	timestamp, err := t.getIntFromArgument(string(arguments[0]))
 	if err != nil {
 		return true, err
 	}
 
 	currentTimeStamp := t.getTimestampHandler()
 	if timestamp+int64(hardforkGracePeriod.Seconds()) < currentTimeStamp {
-		return true, fmt.Errorf("%w out of grace period message", update.ErrIncorrectHardforkMessage)
+		return true, fmt.Errorf("%w timestamp out of grace period message", update.ErrIncorrectHardforkMessage)
```
Comment (SebastianMarian) [inline update/trigger/trigger.go:226]: message timestamp out of grace period

File: update/trigger/trigger.go [hunk 16]
```diff
@@ -106,50 +207,68 @@ func (t *trigger) TriggerReceived(originalPayload []byte, data []byte, pkBytes [
...
-func (t *trigger) getTimestampFromPayload(payload []byte) (int64, error) {
-	splits := bytes.Split(payload, []byte(dataSeparator))
-	if len(splits) < 2 {
-		return 0, fmt.Errorf("%w, not enough tokens found: expected 2, found: %d",
-			update.ErrIncorrectHardforkMessage,
-			len(splits),
-		)
+	currentEpoch := int64(t.epochProvider.MetaEpoch())
+	if currentEpoch-epoch > epochGracePeriod {
+		return true, fmt.Errorf("%w epoch out of grace perdiod", update.ErrIncorrectHardforkMessage)
```
Comment (SebastianMarian) [inline update/trigger/trigger.go:236]: typo perdiod

File: epochStart/notifier/epochStartSubscriptionHandler.go [hunk 17]
```diff
@@ -84,6 +87,26 @@ func (essh *epochStartSubscriptionHandler) NotifyAllPrepare(metaHdr data.HeaderH
...
+	}
+
+	essh.mutEpochStartHandler.Lock()
+	essh.epochFinalizedHandler = append(essh.epochFinalizedHandler, handler)
+	essh.mutEpochStartHandler.Unlock()
+}
+
+// NotifyEpochChangeConfirmed will call all the subscribed clients to notify them that an epoch change is confirmed
+func (essh *epochStartSubscriptionHandler) NotifyEpochChangeConfirmed(epoch uint32) {
+	essh.mutEpochStartHandler.Lock()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/notifier/epochStartSubscriptionHandler.go:103]: done

File: update/mock/epochStartNotifierStub.go (hunk 9 above)
Comment (sasurobert) [reply update/mock/epochStartNotifierStub.go:45]: no need - this is a mock. somewhat implementation - copied from integration/mock package.

File: update/mock/epochStartNotifierStub.go (hunk 10 above)
Comment (sasurobert) [reply update/mock/epochStartNotifierStub.go:59]: no need - this is a mock. somewhat implementation - copied from integration/mock package.

File: update/mock/epochStartNotifierStub.go (hunk 11 above)
Comment (sasurobert) [reply update/mock/epochStartNotifierStub.go:70]: no need - this is a mock. somewhat implementation - copied from integration/mock package.

File: update/trigger/trigger.go (hunk 12 above)
Comment (sasurobert) [reply update/trigger/trigger.go:79]: done

File: update/trigger/trigger.go (hunk 13 above)
Comment (sasurobert) [reply update/trigger/trigger.go:118]: done.

File: update/trigger/trigger.go (hunk 14 above)
Comment (sasurobert) [reply update/trigger/trigger.go:130]: done

File: update/trigger/trigger.go (hunk 15 above)
Comment (sasurobert) [reply update/trigger/trigger.go:226]: done

File: update/trigger/trigger.go (hunk 16 above)
Comment (sasurobert) [reply update/trigger/trigger.go:236]: done

File: update/errors.go [hunk 18]
```diff
@@ -203,8 +203,8 @@ var ErrEmptyChainID = errors.New("empty chain ID")
 // ErrNilArgumentParser signals that nil argument parser was provided
 var ErrNilArgumentParser = errors.New("nil argument parser")
 
-// ErrNilExportHandlerFactory signals that nil export handler factory has been provided
-var ErrNilExportHandlerFactory = errors.New("nil export handler factory")
+// ErrNilExportFacotryHandler signals that nil export handler factory has been provided
```
Comment (SebastianMarian) [inline update/errors.go:206]: Typo facotry
"signals that nil export factory handler" instead "signals that nil export handler factory"

File: update/errors.go [hunk 19]
```diff
@@ -203,8 +203,8 @@ var ErrEmptyChainID = errors.New("empty chain ID")
...
+var ErrNilExportFacotryHandler = errors.New("nil export handler factory")
```
Comment (SebastianMarian) [inline update/errors.go:207]: Typo facotry
"nil export factory handler" instead "nil export handler factory"

File: update/trigger/trigger.go (hunk 2 above)
Comment (iulianpascalau) [reply update/trigger/trigger.go:271]: Removed in the next PR

File: update/mock/epochStartNotifierStub.go (hunk 8 above)
Comment (sasurobert) [reply update/mock/epochStartNotifierStub.go:36]: no need - this is a mock. somewhat implementation - copied from integration/mock package.

File: update/errors.go (hunk 18 above)
Comment (sasurobert) [reply update/errors.go]: done

File: update/errors.go (hunk 19 above)
Comment (sasurobert) [reply update/errors.go]: done

**Pull request merged**